From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Michael Ludwig <michaelludwig@google.com>
Date: Fri, 4 Sep 2026 09:38:19 -0400
Subject: [M150] [gpu] Don't mark GlyphData's atlas generation on a partial
 update

Original change's description:
> [gpu] Don't mark GlyphData's atlas generation on a partial update
>
> If a GlyphData/subrun is split across a flush (i.e. begin > 0 for a
> second call to the same GlyphData), we don't want to mark its atlas
> generation as if it had successfully pushed all of its glyphs into the
> atlas.
>
> The previous assumption was that if we reached the end, then all the
> glyphs were still in the atlas. However, if begin > 0, then the [0,
> begin) glyphs were already used and consumed from a previous flush and
> then are seen as evictable. It is plausible then that the remaining
> [begin, end] glyphs had to overwrite earlier ones in order to fit.
>
> This guards the atlas generation fast path for the scenario when all [0,
> glyphCount] glyphs were added in one go, so they really are guaranteed
> to have the same use/flush token.
>
> Bug: 540016074
> Change-Id: Ibefee9d541cf149c2ce5640ea304aa801c590cab
> Reviewed-on: https://skia-review.googlesource.com/c/skia/+/1350696
> Reviewed-by: Thomas Smith <thomsmit@google.com>
> Commit-Queue: Michael Ludwig <michaelludwig@google.com>

(cherry picked from commit 9b1e5fd08d2cbc9c8933c3538716905c4cf19c5a)

Bug: 561837587,540016074
Change-Id: Ibefee9d541cf149c2ce5640ea304aa801c590cab
Reviewed-on: https://skia-review.googlesource.com/c/skia/+/1362296
Commit-Queue: rubber-stamper@appspot.gserviceaccount.com <rubber-stamper@appspot.gserviceaccount.com>
Auto-Submit: Chrome Cherry Picker <chrome-cherry-picker@chops-service-accounts.iam.gserviceaccount.com>
Bot-Commit: rubber-stamper@appspot.gserviceaccount.com <rubber-stamper@appspot.gserviceaccount.com>

diff --git a/src/gpu/ganesh/text/GlyphData.cpp b/src/gpu/ganesh/text/GlyphData.cpp
index 503ac353a5010839f3c42e6a006b37d245f7a84d..e544e79dea5e95725fa34d477094f431ed2aac90 100644
--- a/src/gpu/ganesh/text/GlyphData.cpp
+++ b/src/gpu/ganesh/text/GlyphData.cpp
@@ -190,7 +190,14 @@ std::tuple<bool, int> GlyphData::regenerateAtlas(int begin,
     if (fAtlasGeneration != currentAtlasGen) {
         SkSpan<const Glyph> glyphSpan = glyphVector.accessBackendGlyphs<Glyph>();
         // Calculate the texture coordinates for the vertexes during first use (fAtlasGeneration
-        // is set to kInvalidAtlasGeneration) or the atlas has changed in subsequent calls.
+        // is set to kInvalidAtlasGeneration) or the atlas has changed in subsequent calls. Always
+        // reset the update, even when begin > 0. When begin > 0, this subrun was split across
+        // flushes and previously used glyphs in [0, begin) have an older token. If those glyphs are
+        // not used in [begin, end], they shouldn't be part of the next bulk update (and because
+        // we won't have updated the whole subrun in one go, we won't set `fAtlasGeneration` to take
+        // the fast path on reuse). If we do reuse the glyphs, we need the bulk update to have been
+        // reset so that the call to addGlyphToBulkAndSetUseToken() sees the first use with the new
+        // token and updates the atlas locator.
         fBulkUseUpdater.reset();
 
         SkBulkGlyphMetricsAndImages metricsAndImages{fTextStrike->strikeSpec()};
@@ -219,8 +226,11 @@ std::tuple<bool, int> GlyphData::regenerateAtlas(int begin,
             glyphsPlacedInAtlas++;
         }
 
-        // Update atlas generation if there are no more glyphs to put in the atlas.
-        if (success && begin + glyphsPlacedInAtlas == glyphVector.glyphCount()) {
+        // Update atlas generation if there are no more glyphs to put in the atlas. We can only do
+        // this if we successfully checked/added the entire glyph vector in one pass. Otherwise, on
+        // a partial update, some of the previous glyphs of the subrun that were already drawn could
+        // have been evicted to make room for these remaining glyphs.
+        if (success && begin == 0 && glyphsPlacedInAtlas == glyphVector.glyphCount()) {
             // Need to get the freshest value of the atlas' generation because
             // updateTextureCoordinates may have changed it.
             fAtlasGeneration = atlasManager->atlasGeneration(maskFormat);
diff --git a/src/gpu/graphite/text/GlyphData.cpp b/src/gpu/graphite/text/GlyphData.cpp
index b8f8fe6aad057a0e6a0fac7c5153cdbc6fbc4163..8b72790a2e2fae985a8efcd701d610968fe2810f 100644
--- a/src/gpu/graphite/text/GlyphData.cpp
+++ b/src/gpu/graphite/text/GlyphData.cpp
@@ -56,7 +56,14 @@ std::tuple<bool, int> GlyphData::regenerateAtlas(int begin,
 
     if (fAtlasGeneration != currentAtlasGen) {
         // Calculate the texture coordinates for the vertexes during first use (fAtlasGeneration
-        // is set to kInvalidAtlasGeneration) or the atlas has changed in subsequent calls.
+        // is set to kInvalidAtlasGeneration) or the atlas has changed in subsequent calls. Always
+        // reset the update, even when begin > 0. When begin > 0, this subrun was split across
+        // flushes and previously used glyphs in [0, begin) have an older token. If those glyphs are
+        // not used in [begin, end], they shouldn't be part of the next bulk update (and because
+        // we won't have updated the whole subrun in one go, we won't set `fAtlasGeneration` to take
+        // the fast path on reuse). If we do reuse the glyphs, we need the bulk update to have been
+        // reset so that the call to addGlyphToBulkAndSetUseToken() sees the first use with the new
+        // token and updates the atlas locator.
         fBulkUseUpdater.reset();
 
         SkBulkGlyphMetricsAndImages metricsAndImages{fTextStrike->strikeSpec()};
@@ -82,8 +89,11 @@ std::tuple<bool, int> GlyphData::regenerateAtlas(int begin,
             glyphsPlacedInAtlas++;
         }
 
-        // Update atlas generation if there are no more glyphs to put in the atlas.
-        if (success && begin + glyphsPlacedInAtlas == glyphVector.glyphCount()) {
+        // Update atlas generation if there are no more glyphs to put in the atlas. We can only do
+        // this if we successfully checked/added the entire glyph vector in one pass. Otherwise, on
+        // a partial update, some of the previous glyphs of the subrun that were already drawn could
+        // have been evicted to make room for these remaining glyphs.
+        if (success && begin == 0 && glyphsPlacedInAtlas == glyphVector.glyphCount()) {
             // Need to get the freshest value of the atlas' generation because
             // updateTextureCoordinates may have changed it.
             fAtlasGeneration = atlasManager->atlasGeneration(maskFormat);
